io/private/las/Vlr.cpp

```
@@ -0,0 +1,412 @@
+            vlr.promisedDataSize };
+        insert(entry);
+        evlrOffset += Evlr::HeaderSize + vlr.promisedDataSize;
+        evlrCount--;
+    }
+}
+
+void VlrCatalog::insert(const VlrCatalog::Entry& entry)
+{
+    std::unique_lock<std::mutex> l(m_mutex);
+
+    m_entries.push_back(entry);
+}
+
+std::vector<char> VlrCatalog::fetch(const std::string& userId, uint16_t recordId) const
+{
+    uint64_t offset = 0;
+    uint32_t length = 0;
+
+    // We don't lock m_entries because we assume that the load has already occurred at the
+    // time you want to fetch.
+    std::vector<char> vlrdata;
+    for (const Entry& e : m_entries)
+        if (e.userId == userId && e.recordId == recordId)
+        {
+            // We don't support VLRs with size > 4GB)
+            if (e.length > (std::numeric_limits<uint32_t>::max)())
+                return vlrdata;
+
+            offset = e.offset;
+            length = (uint32_t)e.length;
+            break;
+        }
+
+    if (length > 0)
+        vlrdata = m_fetch(offset, length);
+    return vlrdata;
+}
+
+std::vector<char> VlrCatalog::fetchWithDescription(const std::string& userId, uint16_t recordId,
+    std::string& outDescrip) const
+{
+    uint64_t offset = 0;
+    uint32_t length = 0;
+
+    // We don't lock m_entries because we assume that the load has already occurred at the
+    // time you want to fetch.
+    std::vector<char> vlrdata;
+    for (const Entry& e : m_entries)
+        if (e.userId == userId && e.recordId == recordId)
+        {
+            // We don't support VLRs with size > 4GB)
+            if (e.length > (std::numeric_limits<uint32_t>::max)())
+                return vlrdata;
+
+            offset = e.offset;
+            length = (uint32_t)e.length;
+            break;
+        }
+    if (length == 0)
+        return vlrdata;
+
+    // Load the data plus the description.
+    const int DescripLen = 32;
+    assert(offset > DescripLen);
```

This should throw an error in case someone gave us a junk VLR definition, which happens.